std/js/lib/Iterator.hx

```diff
@@ -22,11 +22,48 @@
 
 package js.lib;
 
-typedef Iterator<T> = {
+/**
+	abstract over JavaScript iterator objects that enables supports haxe for-in iteration
+**/
+@:forward
+abstract Iterator<T>(IteratorStructure<T>) from IteratorStructure<T> {
+
+	public inline function iterator(): JSIterator<T> {
```

Comment:
This is semantically wrong. It makes `js.lib.Iterator` a Haxe's `Iterable`.
The difference is the former could be iterated one time, while the latter could be iterated multiple times.
So `js.lib.Iterator` should behave like Haxe's `Iterator`, which means fields `next():T` and `hasNext():Bool` should be declared directly on `js.lib.Iterator`. But I don't think it's possible because ES6 Iterator already has `next` with a different signature.
Though, that's pretty convenient to have `iterator()` field on `js.lib.Iterable` :)

Reply:
Yeah, I agree about the semantic issue, ideally we'd just have `next():T` and `hasNext()` on the abstract but I couldn't find a way around the collision with the ES6 next()

Maybe the solution here is to rename `abstract Iterator ` to `abstract Iterable` and use that in Set and Map

Reply:
To resolve the semantic issue, I've disabled iterating over `js.lib.Iterator` (now instead there's a class `HaxeIterator` and you have to manually convert it via `new HaxeIterator(jsIterator)`)

But iterating over Set and Map will still work and is semantically sound

(I tried lots of approaches to try to resolve the semantic issue with iterating over `js.lib.Iterator` however I don't think it's possible to be fully correct here)

Reply:
To enable `js.lib.Iterator` iteration you can do

`using HaxeIterator;`

Which avoids any semantic issues because `js.lib.Iterator` still can't unify with `StdTypes.Iterable`